scripts/dev/upstream_merge/utils/toolchain.py

```diff
@@ -0,0 +1,83 @@
+"""Toolchain detection/build helper: auto-detect existing repo first (ancestor walk),
+ignore nilrt_root hint if already inside a clone; only clone when nothing usable exists."""
+import os
+from .shell_commands import run_command
+from .git_commands import git_clone
+
+BUILD_SCRIPT_REL = os.path.join('scripts', 'pipelines', 'build.toolchain.sh')
+
+def _has_build_script(root):
+    return root and os.path.isfile(os.path.join(root, BUILD_SCRIPT_REL))
+
+def _walk_up_for_build_script(start_dir):
+    cur = os.path.abspath(start_dir)
+    while True:
+        if _has_build_script(cur):
+            return cur
+        parent = os.path.dirname(cur)
+        if parent == cur:  # reached filesystem root
+            return None
+        cur = parent
+
+def detect_or_build_cross_compile(config, script_dir):
+    """Return CROSS_COMPILE prefix, building toolchain/SDK if missing.
+    Order:
+      1. Ancestor of script_dir containing build.toolchain.sh (preferred)
+      2. Provided hint (config.nilrt_root) if ancestor not found
+      3. Clone into expanded hint (or ~/nilrt-sdk if empty) if still missing
+    """
+    gcc_binary = config.get_toolchain_gcc_path()
+    if gcc_binary and os.path.isfile(gcc_binary):
+        print(f"[INFO] Found CROSS_COMPILE: {config.toolchain_prefix}")
+        return config.toolchain_prefix
+    if not config.toolchain_prefix:
+        raise RuntimeError('[ERROR] No toolchain_prefix configured')
```

Comment:
Wouldn't this kill the script if the value is not provided? Since we have a proper default path for where the toolchain should be installed, you should use that path instead of terminating early.